### PR TITLE
add indexing to datasets

### PR DESCRIPTION
**Pull request by davidslater**
Currently, you can add a numeric index to loading a dataset, like this:
```
import logging
import coloredlogs
coloredlogs.install(logging.DEBUG)
import numpy as np
from armory.data import datasets
ds = datasets.mnist("test", shuffle_files=False, index=None)
print(np.hstack([next(ds)[1] for i in range(10)]))
# [2 0 4 8 7 6 0 6 3 1]
ds = datasets.mnist("test", shuffle_files=False, index=[1, 3, 5, 7, 9])
print(np.hstack([next(ds)[1] for i in range(5)]))
# [0 8 6 6 1]
```

I need to check to make sure that:
- [x] this works with config files
- [x] enable handling of simple slices
- [x] update documentation
- [x] update ArmoryDataGenerator size
- [x] Add tests

I'd be happy for any thoughts at this point.

### CODE REVIEW

**Comment by davidslater** [comment]
@ng390 I think it's ready for another review.

File: armory/data/datasets.py [hunk 1]
```diff
@@ -269,6 +269,72 @@ def parse_split_index(split: str):
     return "+".join(output_tokens)
 
 
+def filter_by_index(dataset: "tf.data.Dataset", index: list, dataset_size: int):
```
Comment (ng390) [inline armory/data/datasets.py:272]: Very minor point - if dataset_size is expected to be an int, why is it cast to an int later? Should it be a Union of int and another type?

File: armory/data/datasets.py [hunk 2]
```diff
@@ -269,6 +269,72 @@ def parse_split_index(split: str):
...
+        raise ValueError("index must have at least one value")
+    elif index[0] < 0:
+        raise ValueError("index values must be nonnegative")
+
+    index_tensor = tf.constant(index, dtype=tf.int64)
+
+    def enum_index(i, x):
+        i = tf.expand_dims(i, 0)
+        out, _ = tf.raw_ops.ListDiff(x=i, y=index_tensor, out_idx=tf.int64)
+        return tf.size(out) == 0
```
Comment (ng390) [inline armory/data/datasets.py:291]:
```suggestion
        return tf.equal(tf.size(out), 0)
```

Comment (ng390) [reply armory/data/datasets.py:291]: I think this is the compatibility issue that is causing tf1 pytest to fail -- checking the tf1 and tf2 tests with the change locally now.

Comment (ng390) [reply armory/data/datasets.py:291]: With the change, the tests in tests/test_docker/test_dataset.py pass locally on the tf1, tf2, and pytorch containers

Comment (davidslater) [reply armory/data/datasets.py]: Updated.

File: armory/data/datasets.py (hunk 1 above)
Comment (davidslater) [reply armory/data/datasets.py:272]: It's primarily so that if the input is not an int, it will fail fast.

**Comment by davidslater** [comment]
Okay, pushed change. Any further review?

**Comment by ng390** [comment]
LGTM -- good to drop the WIP label?

**Review by ng390 — APPROVED** [review]

**Pull request merged**